File: binarySearch/FindMin154.java
public class FindMin154{
    public static void main(String[] args) {

    }

    //3 3 1 3
    //1 3 3
    public int findMin(int[] nums) {
        int left=0,right=nums.length-1;
        while(left<right){
            int mid=left+(right-left)/2;
            if(nums[mid] > nums[right]){
                left=mid+1;
            }else if(nums[mid] < nums[right]){
                right=mid;
            }else{
                right--; //和右边界相等,无法判断,只缩减一步
            }
        }
        return nums[left];
    }
}

File: binarySearch/FindMin154.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums)-1
        res = 0
        while right >= 0 and nums[right] == nums[0]:
            right -= 1
        nr = right
        while left <= right:
            mid = left + (right - left)//2
            if nums[mid] > nums[nr]:
                left = mid + 1
            else:
                res = mid
                right = mid - 1
        return nums[res]